fix: Sends failing on UTP adapter when queue is full

simon-lemay-unity opened the pull request:
Or close to full. There were two issues here.

First, the payload capacity of the fragmentation pipeline didn't account for the entire size of the packet. Unfortunately, payload capacity in the context of the fragmentation pipeline is a bit of a misnomer since it also includes headers and other such overhead. We were setting it to the size of a full send queue, and so when you added UTP headers, we ended up with a packet too large for the pipeline. Bumping up the value a bit solves this issue.

Second, when sending a single message that was within 4 bytes of the send queue capacity, we'd silently fail to send it. The reason was we didn't account for the extra overhead taken by the data length in the send queue (which requires 4 bytes). Thus the send code would think the message could fit in the queue when it couldn't. Fixing this simply required accounting for the 4 extra bytes in the send logic.

Two tests were also added to test sends when the send queue is full (either from one big send or from multiple smaller ones in a row). Both tests fail without the fixes in the adapter.

## Changelog

### com.unity.netcode.adapter.utp

* Fixed: Sends failing when send queue is full or close to full.

## Testing and Documentation

* Includes unit tests.
* No documentation changes or additions were necessary.

simon-lemay-unity committed:
fix: Sends failing on UTP adapter when queue is full

Or close to full. There were two issues here.

First, the payload capacity of the fragmentation pipeline didn't account
for the entire size of the packet. Unfortunately, payload capacity in
the context of the fragmentation pipeline is a bit of a misnomer since
it also includes headers and other such overhead. We were setting it to
the size of a full send queue, and so when you added UTP headers, we
ended up with a packet too large for the pipeline. Bumping up the value
a bit solves this issue.

Second, when sending a single message that was within 4 bytes of the
send queue capacity, we'd silently fail to send it. The reason was we
didn't account for the extra overhead taken by the data length in the
send queue (which requires 4 bytes). Thus the send code would think the
message could fit in the queue when it couldn't. Fixing this simply
required accounting for the 4 extra bytes in the send logic.

Two tests were also added to test sends when the send queue is full
(either from one big send or from multiple smaller ones in a row). Both
tests fail without the fixes in the adapter.

simon-lemay-unity committed:
Cleaner fix for the send queue being full scenario

This one depends on a fix in UTP that's still being reviewed, though.

simon-lemay-unity commented:
I've updated the fix to avoid having to use extra bytes in the fragmentation pipeline capacity to get a full send queue to work correctly. It was hacky and was actually papering over an actual issue in UTP. The fix for the UTP issue is in review (tentatively scheduled for 1.0.0-pre.7). If we're going to merge this PR before we depend on that new version, we'll want to disable the new tests.

Note that this fix is still just adding extra bytes to the fragmentation pipeline payload capacity, but now the extra bytes are only there to account for overhead caused by the UTP adapter itself (by the send queue, to be more specific). And it's also exactly how many extra bytes are required so it doesn't feel like some magic number.

Cosmin-B commented on com.unity.netcode.adapter.utp/Runtime/UnityTransport.cs:
```diff
@@ -599,7 +602,8 @@ public override void Send(ulong clientId, ArraySegment<byte> payload, NetworkDel
                 // If we are in here data exceeded remaining queue size. This should not happen under normal operation.
-                if (payload.Count > queue.Size)
+                // Extra 4 bytes here are to account for the data size that will need to be stored in the queue too.
+                if (payload.Count + 4 > queue.Size)
```
Why are we just not using "size" from above, which includes the extra 4 bytes for the data + 1 byte for the batch/unbatched (which we are still writing) and the payload.Count

simon-lemay-unity replied:
Because that check is to decide if the data will fit in the send queue or not. We don't store the byte for batched/unbatched in the send queue, it's only prepended when we're about to send the data.

But looking at this code a bit more, I don't think the logic is correct here. Currently it's sending the data immediately even if there is something in the send queue. Say if messages A and B are in the queue and we're trying to add C but it doesn't fit (that check fails), we'll immediately send C and then later on send A and B batched. If all messages have a reliable sequenced delivery method, we'll deliver them in the order C, A, and B. That's not correct.

Instead, the logic should be that if adding a message fails, it should always flush the queue, then try adding it again, and if that fails again, just send it immediately. That removes the need to account for these 4 extra bytes here since it would be done implicitly when trying to add the message. I'll edit my PR with that fix.

Cosmin-B replied:
Well, it depends, what you are describing, in practice only happens once, which is the Initial Payload, hence we send that immediately as that is hopefully the only packet that is bigger than the entire preallocated size of the SendQueue. (you can also notice that Debug.Assert there which is specifically looking for the ReliableFragmentSequenced delivery method, and in NGO that Is only used for Initial Payload)

My implementation above does not handle the ordering correctly for that specific C/A and B use-case only for the initial payload, which is not really a big deal. But we can modify it to be correctly ordered and it should be a quick fix.

simon-lemay-unity committed:
Fix possible re-ordering issue for large messages

simon-lemay-unity committed:
Reword the warning when sending immediately

Events can't be larger than m_SendQueueBatchSize. UTP will drop them and
log an error in the current configuration. Reword the warning so that we
don't imply that it's been sent, only that we're trying to send it. If
it fails, there will be an error in the log following it, making it
clear to the user what happened.

simon-lemay-unity committed:
Update UTP dependency to 1.0.0-pre.7

andrews-unity committed:
Merge branch 'develop' into fix/utp-adapter-filled-queue

andrews-unity committed:
fix: Fixing validation error

andrews-unity committed:
Merge branch 'develop' into fix/utp-adapter-filled-queue

andrews-unity committed:
Merge branch 'develop' into fix/utp-adapter-filled-queue

andrews-unity committed:
Merge branch 'develop' into fix/utp-adapter-filled-queue
